crates/ty_python_semantic/src/types.rs

```diff
@@ -284,6 +284,81 @@ fn class_lookup_cycle_initial<'db>(
     Place::bound(Type::Never).into()
 }
 
+pub(crate) trait VarianceInferable<'db>: Sized {
+    fn variance_of(self, db: &'db dyn Db, type_var: TypeVarInstance<'db>) -> TypeVarVariance;
+
+    fn with_polarity(self, polarity: TypeVarVariance) -> WithPolarity<Self> {
+        WithPolarity {
+            variance_inferable: self,
+            polarity,
+        }
+    }
+}
+
+pub(crate) struct WithPolarity<T> {
+    variance_inferable: T,
+    polarity: TypeVarVariance,
+}
+
+impl<'db, T> VarianceInferable<'db> for WithPolarity<T>
+where
+    T: VarianceInferable<'db>,
+{
+    // Based on the variance composition/transformation operator in
+    // https://people.cs.umass.edu/~yannis/variance-extended2011.pdf, page 5
+    //
+    // While their operation has compose(invariant, bivariant) = invariant, we
+    // instead have it evalaute to bivariant. This is a valid choice, as
+    // discussed on that same page, where type equality is semantic rather than
+    // syntactic. To see that this holds for our setting consider the type
+    // ```python
+    // type ConstantInt[T] = int
+    // ```
+    // We would say `ConstantInt[str]` = ConstantInt[float], so we qualify as
+    // using semantic equivalence.
+    fn variance_of(self, db: &'db dyn Db, type_var: TypeVarInstance<'db>) -> TypeVarVariance {
+        let WithPolarity {
+            variance_inferable,
+            polarity,
+        } = self;
+        match polarity {
+            TypeVarVariance::Covariant => variance_inferable.variance_of(db, type_var),
+            TypeVarVariance::Contravariant => variance_inferable.variance_of(db, type_var).flip(),
+            TypeVarVariance::Bivariant => TypeVarVariance::Bivariant,
+            TypeVarVariance::Invariant => {
+                if let TypeVarVariance::Bivariant = variance_inferable.variance_of(db, type_var) {
```

Comment:
nit: `TypeVarVariance` is a non-data-carrying enum that implements `Eq`, so you can just use an equality comparison here